### ISSUE TITLE
Incorrect results of DTensor.mean with uneven sharding

### ISSUE THREAD

**Issue by ppwwyyxx** [pytorch/pytorch#162692]
### 🐛 Describe the bug

```python
from torch.distributed._tensor import Shard, distribute_tensor, init_device_mesh
import torch
from torch import distributed as dist

if __name__ == '__main__':
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    torch.cuda.set_device(rank)

    tensor = torch.arange(12).reshape(-1, 4).float().cuda()
    print("Truth=", tensor.mean())

    mesh = init_device_mesh('cuda', (2,))
    dt = distribute_tensor(tensor, device_mesh=mesh, placements=[Shard(0)])

    mean = dt.mean()
    full = mean.full_tensor()
    print(rank, mean, full)
```
The above code attempts to compute mean of a 3x4 DTensor. The result does not account for the uneven sharding and is incorrect.

In addition, if input is 1x4, it results in NaN.

Run it with `torchrun --nproc-per-node=2 a.py`

### Versions

torch2.8

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @dcci @tianyu-l @XilunWu

**Comment by wanchaol** [comment]
> I propose we fix this by just banning Partial(avg) when it's imbalanced. It is hard for me to see another way to do it, if you want to do mean locally you have:
> 
> ```
> aten.mean.default (LocalTensor(
>   0: tensor([[0., 1., 2., 3.],
>         [4., 5., 6., 7.]]),
>   1: tensor([[ 8.,  9., 10., 11.],
>         [ 0.,  0.,  0.,  0.]])
> ),) {}
> ```
> 
> But there is no simple deferred collective you can run to fix this, since you need to weight 1 differently in the final mean computation.

@ezyang I agree on the approach!

**Comment by ezyang** [comment]
```
diff --git a/torch/distributed/tensor/_ops/_math_ops.py b/torch/distributed/tensor/_ops/_math_ops.py
index 1e6eb40939e..14d8c5dd21a 100644
--- a/torch/distributed/tensor/_ops/_math_ops.py
+++ b/torch/distributed/tensor/_ops/_math_ops.py
@@ -268,7 +268,17 @@ def common_reduction_strategy(
     reduction_strategy = OpStrategy([])
 
     for op_spec in input_strategy.strategies:
-        if not reduction_linear:
+        padded_reduction = False
+        output_spec = op_spec.output_specs
+        local_shape = list(output_spec.tensor_meta.shape)
+        for mesh_d, p in enumerate(output_spec.placements):
+            if isinstance(p, Shard) and p.dim in reduce_dims:
+                if local_shape[p.dim] % output_spec.mesh.size(mesh_d) != 0:
+                    padded_reduction = True
+                    break
+                local_shape[p.dim] //= output_spec.mesh.size(mesh_d)
+
+        if not reduction_linear or padded_reduction:
             # input placements for this strategy should clear out pending sum and sharding
             # on the reduction dimension
             input_placements = replicate_reduction_dims(
@@ -307,6 +317,10 @@ LINEAR_REDUCTION_OP_MAP = {
     aten.all.dim: "sum",
     aten.sum.default: "sum",
     aten.sum.dim_IntList: "sum",
+    aten.any.default: "sum",
+    aten.any.dim: "sum",
+    aten.any.out: "sum",
+    # These are only valid when there is no padding
     aten.prod.default: "product",
     aten.prod.dim_int: "product",
     aten.prod.int_out: "product",
@@ -319,9 +333,6 @@ LINEAR_REDUCTION_OP_MAP = {
     aten.min.default: "min",
     aten.min.dim: "min",
     aten.min.out: "min",
-    aten.any.default: "sum",
-    aten.any.dim: "sum",
-    aten.any.out: "sum",
     aten.amax.default: "max",
     aten.amax.out: "max",
     aten.amin.default: "min",
```

**Comment by SherlockNoMad** [comment]
> I feel like there ought to be a better way to test if padding occurred but I didn't see anything obvious on OpSpec

I implemented is_tensor_evenly_shardable_on_dim() as util func.
